manage/config-dialog.js

```diff
@@ -286,7 +314,26 @@ function configDialog(style) {
   }
 
   function updateVarOnChange() {
-    this.va.value = this.type !== 'checkbox' ? this.value : this.checked ? '1' : '0';
+    if (this.type === 'number') {
+      this.value = this.va.value = clampValue(this.value, this.va.range);
+    } else if (this.type === 'range') {
+      $('.current-value', this.closest('.config-range')).textContent = this.va.value + (this.va.units || '');
+      this.va.value = parseFloat(this.value);
+    } else {
+      this.va.value = this.type !== 'checkbox' ? this.value : this.checked ? '1' : '0';
+    }
+  }
+
+  // Clamp input[type=number] to a valid range
+  function clampValue(value, [min = 0, max = 100, step]) {
+    if (value < min) {
+      return min;
+    } else if (value > max) {
+      return max;
+    }
+    const remainder = value % (step || 1);
```

Comment:
This won't work with floats because of the precision. If we really need this, we have to upscale the value to an integer then downscale them after the calculation.

Reply:
I tested this using a step of 0.5...

```
1.9 % 0.5
// => 0.3999999999999999
// but
1.9 - (1.9 % 0.5)
// => 1.5

So the precision should cancel out...

Reply:
0.5 can be saved as float without losing precision. Try `1 - 1 % 0.1`, because 0.1 float number is actually slightly larger than actual 0.1, the result of 1 % 0.1 is slightly smaller than 0.1, then 1 - (1 % 0.1) becomes 0.9.

Reply:
Oh yeah, good example.... so maybe just leave it out then?

Reply:
Detect the precision of `step` and round the result to it?
Something like stepPrecision = stepStr.match(/\.(\d+?[1-9])0*\b|$/)[1].length;
N.B. A reliable maximum is 15 digits for IEEE754 floats used in JS including all the digits, not just the fractional part.

Reply:
We don't have to round the value for users IMHO. If the author made some validations and the user don't care about the error, we can accept it as well.

The main problem is how to tell the user that the input is invalid. I think we can:

1. Add some styles to `input:invalid`.
2. Call `input.reportValidity()` on `change` event to show the error message.

Reply:
I changed the method of adjusting the value to the nearest step below the value... not sure what to call it, clamp to step? Anyway, it should work now; it might still need the `.toFixed()` function for cleanup...